### ISSUE TITLE
[Phase 3] Training Loop

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/diffusion-boltzmann-sampler#5]
## Objective

Implement denoising score matching training loop with proper loss function and logging.

## Deliverables

- [ ] `ScoreMatchingLoss` computing denoising objective
- [ ] `Trainer` class with training loop and checkpointing
- [ ] `/training/start` POST endpoint to initiate training
- [ ] `/training/status` GET endpoint to check progress
- [ ] Training metrics logging

## Files to Create

| File | Description |
|------|-------------|
| `backend/ml/training/trainer.py` | Training loop implementation |
| `backend/ml/training/losses.py` | Score matching loss |
| `backend/api/routes/training.py` | Training API endpoints |
| `backend/tests/test_training.py` | Training tests |

## Core Implementation

```python
# backend/ml/training/losses.py
import torch

def denoising_score_matching_loss(
    model: torch.nn.Module,
    x_0: torch.Tensor,
    diffusion: "DiffusionProcess"
) -> torch.Tensor:
    """DSM loss: E_t E_{x_t|x_0} ||s_θ(x_t,t) - ∇log p(x_t|x_0)||²

    For Gaussian noising: ∇log p(x_t|x_0) = -noise/σ_t
    """
    batch_size = x_0.shape[0]
    device = x_0.device

    # Sample time uniformly
    t = torch.rand(batch_size, device=device)

    # Get noisy samples
    x_t, noise = diffusion.forward(x_0, t)
    _, sigma_t = diffusion.noise_level(t)

    # Target score
    target = -noise / (sigma_t[:, None, None, None] + 1e-8)

    # Predicted score
    pred = model(x_t, t)

    # Weighted MSE (weight by σ²)
    loss = ((pred - target) ** 2).mean()
    return loss
```

```python
# backend/ml/training/trainer.py
class Trainer:
    def __init__(self, model, diffusion, lr=1e-3, device="cpu"):
        self.model = model.to(device)
        self.diffusion = diffusion
        self.device = device
        self.optimizer = torch.optim.Adam(model.parameters(), lr=lr)
        self.history = {"loss": []}

    def train_epoch(self, dataloader):
        self.model.train()
        total_loss = 0
        for batch in dataloader:
            batch = batch.to(self.device)
            self.optimizer.zero_grad()
            loss = denoising_score_matching_loss(self.model, batch, self.diffusion)
            loss.backward()
            torch.nn.utils.clip_grad_norm_(self.model.parameters(), 1.0)
            self.optimizer.step()
            total_loss += loss.item()
        avg_loss = total_loss / len(dataloader)
        self.history["loss"].append(avg_loss)
        return avg_loss

    def save_checkpoint(self, path: str):
        torch.save({
            "model": self.model.state_dict(),
            "optimizer": self.optimizer.state_dict(),
            "history": self.history
        }, path)
```

## Verification

```python
def test_loss_decreases():
    """Training should reduce loss."""
    model = ScoreNetwork(in_channels=1, base_channels=16)
    trainer = Trainer(model, DiffusionProcess())

    # Generate random "data"
    data = torch.randn(100, 1, 16, 16)
    loader = DataLoader(data, batch_size=16)

    loss_0 = trainer.train_epoch(loader)
    for _ in range(10):
        trainer.train_epoch(loader)
    loss_10 = trainer.history["loss"][-1]

    assert loss_10 < loss_0, "Loss should decrease"
```

## Definition of Done

- Loss decreases over training epochs
- Checkpoints save and load correctly
- API endpoints return training status
- Gradient clipping prevents explosions

## Dependencies

- Phase 2: Score Network

## Related

Meta issue: #1
